source/simulator/solver.cc

```diff
@@ -661,26 +661,29 @@ namespace aspect
         // linearized_stokes_variables has a different
         // layout than current_linearization_point, which also contains all the
         // other solution variables.
-        if (parameters.nonlinear_solver != NonlinearSolver::Newton_Stokes)
+
+        if (!assemble_newton_stokes_system)
           {
             linearized_stokes_initial_guess.block (block_vel) = current_linearization_point.block (block_vel);
-
             linearized_stokes_initial_guess.block (block_p) = current_linearization_point.block (block_p);
+
             denormalize_pressure (this->last_pressure_normalization_adjustment,
                                   linearized_stokes_initial_guess,
                                   current_linearization_point);
           }
         else
           {
             /**
-             * The Newton solver solves for updates to variables, for which we have no good initial guesses.
-             * Therefore, initialize the solution vector for the solver to zero.
-             * TODO: I think that for the first iteration we could actually use a non-zero inital guess.
+             * The Newton solver solves for updates to variables, for which our best guess is zero when
+             * the it isn't the first nonlinear iteration. When it is the first nonlinear iteration, we
+             * have to assemble the full (non-defect correction) Picard, to get the boundary conditions
+             * right in combination with being able to use the initial guess optimally.
```

Comment:
Hmm, I guess I should rephrase it then. What I try to explain here is that for the Newton solver the initial guess has to be always zero (which is done in the lines below here). This is something you might not expect, because normally you set in the first nonlinear iteration of the newton solver the initial guess to the expected solution (current liniasation point for example). We don't do that here because we always use a full Picard iteration for that nonlinear iteration. Does this make more sense?